File: code/python/525.findMaxLength.py
class Solution:
    def findMaxLength(self, nums: List[int]) -> int:
        count = 0
        m = dict()
        mx = 0
        m[0] = -1
        for i in range(len(nums)):
            if nums[i] == 0:
                count -= 1
            elif nums[i] == 1:
                count += 1
            if m.get(count) is not None:
                mx = max(mx, i - m.get(count))
            else:
                m[count] = i
        return mx

File: code/cpp/525.findMaxLength.cpp
#include <vector>
#include <unordered_map>

class Solution
{
public:
    int findMaxLength(vector<int> &nums)
    {
        map<int, int> m{{0, -1}};
        int sum = 0;
        int ans = 0;
        for (int i = 0; i < nums.size(); ++i)
        {
            sum += !nums[i] ? -1 : 1;
            if (m.count(sum))
                ans = max(ans, i - m[sum]);
            else
                m[sum] = i;
        }
        return ans;
    }
};